### ISSUE TITLE
No matching package when using ^ with pre-1.0 versions

### ISSUE THREAD

**Issue by andig** [composer/composer#5086]
With the following composer.json:

    {
        "name": "test/test",
        "authors": [
            {
                "name": "andig",
                "email": "test@test.de"
            }
        ],
        "require": {
            "react/socket-client": "^0.4",
            "react/http": "dev-master"
        }
    }

When I run this command:

```
composer install
composer require -vvv php-pm/php-pm:dev-master
```

I get this output:

	pi@prodpi ~/test2 $ composer require -vvv php-pm/php-pm:dev-master
	Reading ./composer.json
	Loading config file /home/pi/.composer/config.json
	Loading config file /home/pi/.composer/auth.json
	Loading config file ./composer.json
	Checking CA file /etc/ssl/certs/ca-certificates.crt
	Executing command (/home/pi/test2): git branch --no-color --no-abbrev -v
	Failed to initialize global composer: Composer could not find the config file: /home/pi/.composer/composer.json
	To initialize a project, please create a composer.json file as described in the https://getcomposer.org/ "Getting Started" section
	Reading /home/pi/test2/vendor/composer/installed.json
	./composer.json has been updated
	Reading ./composer.json
	Loading config file /home/pi/.composer/config.json
	Loading config file /home/pi/.composer/auth.json
	Loading config file ./composer.json
	Executing command (/home/pi/test2): git branch --no-color --no-abbrev -v
	Failed to initialize global composer: Composer could not find the config file: /home/pi/.composer/composer.json
	To initialize a project, please create a composer.json file as described in the https://getcomposer.org/ "Getting Started" section
	Reading /home/pi/test2/vendor/composer/installed.json
	Reading ./composer.lock
	Loading composer repositories with package information
	Downloading https://packagist.org/packages.json
	Writing /home/pi/.composer/cache/repo/https---packagist.org/packages.json into cache
	Updating dependencies (including require-dev)
	Reading /home/pi/.composer/cache/repo/https---packagist.org/p-provider-2013.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/p-provider-2014.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/p-provider-2015.json from cache
	Downloading http://packagist.org/p/provider-2015-04%24613b05fce10207a1c94baf39ecbf968560f853292b6b68f34b96fa4696658f86.json
	Writing /home/pi/.composer/cache/repo/https---packagist.org/p-provider-2015-04.json into cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/p-provider-2015-07.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/p-provider-2015-10.json from cache
	Downloading http://packagist.org/p/provider-2016-01%24c3a38e6d71d9ee37a49e4ea3fed55a08830cba62142c3e8192ea9f86eb2fc8db.json
	Writing /home/pi/.composer/cache/repo/https---packagist.org/p-provider-2016-01.json into cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/p-provider-archived.json from cache
	Downloading http://packagist.org/p/provider-latest%240399eaaa23be5dd3ad83bee459e79fd672e77cc88263af05bc355515b21a6848.json
	Writing /home/pi/.composer/cache/repo/https---packagist.org/p-provider-latest.json into cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-react$promise.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-evenement$evenement.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-react$stream.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-react$event-loop.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-react$socket.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-react$cache.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-react$dns.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-react$socket-client.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-psr$http-message.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-guzzlehttp$psr7.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-react$http.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-php-pm$php-pm.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-mkraemer$react-pcntl.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-symfony$console.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-symfony$process.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-symfony$debug.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-monolog$monolog.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-symfony$polyfill-mbstring.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-psr$log.json from cache
	Reading /home/pi/.composer/cache/repo/https---packagist.org/provider-symfony$http-kernel.json from cache
	Your requirements could not be resolved to an installable set of packages.

	  Problem 1
	    - Installation request for php-pm/php-pm dev-master -> satisfiable by php-pm/php-pm[dev-master].
	    - php-pm/php-pm dev-master requires react/socket-client ^0.5.0 -> no matching package found.

	Potential causes:
	 - A typo in the package name
	 - The package is not available in a stable-enough version according to your minimum-stability setting
	   see <https://getcomposer.org/doc/04-schema.md#minimum-stability> for more details.

	Read <https://getcomposer.org/doc/articles/troubleshooting.md> for further common problems.

	Installation failed, reverting ./composer.json to its original content.

**And I expected this to happen:**

Instead of installing, I see an error `react/socket-client ^0.5.0 -> no matching package found`. However, simply running `composer require react/socket-client:^0.5` works fine.

As ^ operator forbids loading 0.5 if 0.4 is already installed I was expecting a version constraint mismatch, bot a "not found". As 0.5.0 for this package had only just been released it kept chasing me up the wrong tree of caches etc.

Wrong expectation on composer error messages?

**Comment by alcohol** [comment]
It's very hard to provide elaborate error messages from a dependency solver. Any number of reasons could be interpreted from the same problem on a solver level. I'm afraid there isn't much we can improve here.

**Comment by andig** [comment]
> It's very hard to provide elaborate error messages from a dependency solver. 

I was expecting that the solver can distinguish between something that does not exist (on packagist) or is not compatible with other constraints. If that's not the case please feel free to close.

**Comment by Seldaek** [comment]
That's due to an optimization we do, since we know that your root file MUST be installed, and it requires ^0.4, we eliminate any version that is not compatible with that reuqirement, so the solver doesn't analyze more than needed, but it also doesn't know that 0.5 exists at all. Solved by #5120

**Comment by andig** [comment]
Excellent and thank you so much for composer!
